Re-land: Race condition between module script execution and FrameLoader's document completion check.

basuke opened the pull request:
#### 85e09b241c129c868e96cdc02c5368cdd61abb96
<pre>
Re-land: Race condition between module script execution and FrameLoader&apos;s document completion check.
<a href="https://bugs.webkit.org/show_bug.cgi?id=297939">https://bugs.webkit.org/show_bug.cgi?id=297939</a>
<a href="https://rdar.apple.com/162211435">rdar://162211435</a>

Reviewed by Ryosuke Niwa.

Module scripts that create microtasks were seeing inconsistent document.readyState values, the
module script itself saw &quot;interactive&quot; but Promise callbacks saw &quot;complete&quot;. This happened
because FrameLoader::finishedParsing() was being called during microtask checkpoint execution.

The fix detects when attemptToRunDeferredScriptsAndEnd() is called during a microtask checkpoint
and defers the end() call using the regular task queue, ensuring finishedParsing() runs after
all microtasks complete.

Because this changes the execution timing of some test code, a regression can be seen on iOS&apos;s
scrolling behavior. When WebPage gets visibleContentRectUpdateInfo from UIProcess, added
validation to check the existance of local change after last CommitLayerTree is sent.

This was originally landed on 299282@main and reverted on 301215@main.

Tests: http/wpt/dom/inline-module-script-timing.html
       http/wpt/dom/inline-module-script-with-imports.html
       imported/w3c/web-platform-tests/css/css-scroll-anchoring/shadow-dom-subscroller.html

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/dom/inline-module-script-timing-expected.txt: Added.
* LayoutTests/http/wpt/dom/inline-module-script-timing.html: Added.
* LayoutTests/http/wpt/dom/inline-module-script-with-imports-expected.txt: Added.
* LayoutTests/http/wpt/dom/inline-module-script-with-imports.html: Added.
* LayoutTests/http/wpt/dom/resources/delayed-module.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/microtasks/evaluation-order-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-origin-traversal-redirect-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/html/parser/HTMLDocumentParser.cpp:
(WebCore::HTMLDocumentParser::notifyFinished):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::applyScrollUpdate):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::willCommitLayerTree):
(WebKit::WebPage::shouldIgnoreScrollPositionUpdate const):
(WebKit::WebPage::markPendingLocalScrollPositionChange):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::markPendingLocalScrollPositionChange):
(WebKit::WebPage::shouldIgnoreScrollPositionUpdate const):
(WebKit::WebPage::updateVisibleContentRects):

Canonical link: <a href="https://commits.webkit.org/301345@main">https://commits.webkit.org/301345@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2854583cebc9d157b847bf6f68d639a9c2a0ac27

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125670 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45333 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36082 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132534 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77554 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/043c493d-a265-4c8c-9948-8fda5f5f3750) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127541 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46018 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53892 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95741 "Found 1 new test failure: webgl/2.0.y/conformance/textures/misc/video-rotation.html (failure)") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63860 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/da705d43-1579-4a77-bd44-fead7a5af875) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128618 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36787 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112381 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76235 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ddb1c1bb-838b-4b27-9145-ec0dba17c8b8) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35689 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30563 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76004 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106565 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30780 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135208 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52463 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40225 "Found 1 new test failure: fast/parser/entity-comment-in-textarea.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104208 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52910 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108592 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103935 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26469 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49300 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27607 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49690 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52358 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58162 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51705 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55057 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53402 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa reviewed (approved):
Looks good outside the scroll code change. @smfr : could you review the scroll code change?

basuke committed:
Re-land: Race condition between module script execution and FrameLoader's document completion check.
https://bugs.webkit.org/show_bug.cgi?id=297939
rdar://162211435

Reviewed by Ryosuke Niwa.

Module scripts that create microtasks were seeing inconsistent document.readyState values, the
module script itself saw "interactive" but Promise callbacks saw "complete". This happened
because FrameLoader::finishedParsing() was being called during microtask checkpoint execution.

The fix detects when attemptToRunDeferredScriptsAndEnd() is called during a microtask checkpoint
and defers the end() call using the regular task queue, ensuring finishedParsing() runs after
all microtasks complete.

Because this changes the execution timing of some test code, a regression can be seen on iOS's
scrolling behavior. When WebPage gets visibleContentRectUpdateInfo from UIProcess, added
validation to check the existance of local change after last CommitLayerTree is sent.

This was originally landed on 299282@main and reverted on 301215@main.

Tests: http/wpt/dom/inline-module-script-timing.html
       http/wpt/dom/inline-module-script-with-imports.html
       imported/w3c/web-platform-tests/css/css-scroll-anchoring/shadow-dom-subscroller.html

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/dom/inline-module-script-timing-expected.txt: Added.
* LayoutTests/http/wpt/dom/inline-module-script-timing.html: Added.
* LayoutTests/http/wpt/dom/inline-module-script-with-imports-expected.txt: Added.
* LayoutTests/http/wpt/dom/inline-module-script-with-imports.html: Added.
* LayoutTests/http/wpt/dom/resources/delayed-module.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/microtasks/evaluation-order-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-origin-traversal-redirect-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/html/parser/HTMLDocumentParser.cpp:
(WebCore::HTMLDocumentParser::notifyFinished):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::applyScrollUpdate):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::willCommitLayerTree):
(WebKit::WebPage::shouldIgnoreScrollPositionUpdate const):
(WebKit::WebPage::markPendingLocalScrollPositionChange):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::markPendingLocalScrollPositionChange):
(WebKit::WebPage::shouldIgnoreScrollPositionUpdate const):
(WebKit::WebPage::updateVisibleContentRects):

Canonical link: https://commits.webkit.org/301345@main
